routemaster/models/queue.rb

```diff
@@ -1,130 +1,189 @@
 require 'routemaster/models'
-require 'routemaster/models/message'
-require 'routemaster/mixins/log'
 require 'routemaster/mixins/redis'
-require 'routemaster/services/codec'
+require 'routemaster/mixins/log'
+require 'routemaster/mixins/log_exception'
+require 'routemaster/models/job'
 
 module Routemaster
   module Models
-    # Abstraction for a queue.
-    # Takes a Subscriber and allows to push Messages in and pull them out in
-    # order.
+    # A collection of `Job`s, scheduled or queued, with deduplication and safe
+    # mutation (push/pop).
     class Queue
-      include Mixins::Log
       include Mixins::Redis
+      include Mixins::Log
+      include Mixins::LogException
 
-      attr_reader :subscriber
+      attr_reader :name
 
-      def initialize(subscriber)
-        @subscriber = subscriber
+      class Retry < StandardError
+        attr_reader :delay
+        def initialize(delay)
+          @delay = delay
+        end
       end
 
-      def pop
-        uid, payload = _redis_lua_run(
-          'pop',
-          keys: [_new_uuids_key, _pending_uuids_key, _payloads_key],
-          argv: [Routemaster.now])
 
-        return if uid.nil?
+      def initialize(name:)
+        @name = name
+      end
 
-        if payload.nil?
-          _log.error { "missing payload for message #{uid} in queue #{@subscriber.name}" }
-          return
+
+      # Returns a list of all queued jobs.
+      # Only use this for testing.
+      def jobs
+        [].tap do |result|
+          _redis.lrange(_queue_key, 0, -1).each do |raw|
+            result << Job.load(raw)
+          end
+          _redis.zrange(_scheduled_key, 0, -1, with_scores: true).each do |raw,score|
+            result << Job.load(raw, run_at: score.to_i)
+          end
         end
+      end
 
-        Services::Codec.new.load(payload, uid)
+      
+      # Returns the job currently being run by `worker_id`.
+      # Empty if no jobs are running, or the worker isn't known.
+      def running_jobs(worker_id)
+        raw = _redis.lrange(_pending_key(worker_id), 0, 0)
+        return [] unless raw
+        raw.map { |x| Job.load(x) }
       end
 
-      def peek
-        uid, payload = _redis_lua_run(
-          'peek',
-          keys: [_new_uuids_key, _payloads_key], 
-          argv: [])
 
-        return if uid.nil?
+      # Count all jobs; or all jobs before the deadline, if specified.
+      def length(deadline:nil, scheduled: true, instant: true)
+        deadline ||= '+inf'
```

Comment:
That would be a leaky abstraction? As the caller would "see" `+inf` which is a Redis specific implementation detail.
I could add an assertion; the intent is that the `deadline` argument is either `nil` or a timestamp.